bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceUse.java

```diff
@@ -28,22 +37,68 @@
  * @ThreadSafe
  */
 public class ServiceUse<S> {
+	/** BundleContext associated with this service use */
+	final BundleContextImpl context;
 	/** ServiceDescription of the registered service */
 	final ServiceRegistrationImpl<S> registration;
+	final Debug debug;
 
 	/** bundle's use count for this service */
 	/* @GuardedBy("this") */
 	private int useCount;
 
+	/** lock for this service */
+	private final ReentrantLock lock = new ReentrantLockExt();
+
+	private Duration lockTimeout = Duration.ofSeconds(10);
+
 	/**
 	 * Constructs a service use encapsulating the service object.
 	 *
 	 * @param   context bundle getting the service
 	 * @param   registration ServiceRegistration of the service
 	 */
 	ServiceUse(BundleContextImpl context, ServiceRegistrationImpl<S> registration) {
-		this.useCount = 0;
+		this.context = context;
 		this.registration = registration;
+		this.useCount = 0;
+		this.debug = context.getContainer().getConfiguration().getDebug();
+	}
+
+	void runWithLock(Runnable runnable) {
+		callWithLock(Executors.callable(runnable));
+	}
+
+	<T> T callWithLock(Callable<T> callable) {
+		try {
+			if (lock.tryLock(getLockTimeout().toMillis(), TimeUnit.MILLISECONDS)) {
+				try {
+					return callable.call();
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt(); // To appease sonar
+					return null;
+				} catch (Exception e) {
+					throw new RuntimeException("Callable threw exception", e); //$NON-NLS-1$
+				} finally {
+					lock.unlock();
+				}
+			}
+			throw new ServiceException(
+					"Failed to acquire lock within try period, Lock, id:" + System.identityHashCode(lock) + ", " //$NON-NLS-1$ //$NON-NLS-2$
+							+ lock.toString(),
+					ServiceException.DEADLOCK);
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt(); // To appease sonar
+			return null;
+		}
+	}
+
+	Duration getLockTimeout() {
+		return lockTimeout;
+	}
+
+	ReentrantLock getLock() {
+		return lock;
 	}
```

Comment:
```suggestion
	interface Locked extends AutoCloseable {
		@Override
		void close();
	}

	Locked lock() throws InterruptedException {
		if (lock.tryLock(lockTimeout, TimeUnit.MILLISECONDS)) {
			if (debug.DEBUG_SERVICES) {
				Debug.println('[' + Thread.currentThread().getName() + "] getServiceLock[" //$NON-NLS-1$
						+ context.getBundle() + "](" + this + "), id:" + System.identityHashCode(lock) //$NON-NLS-1$ //$NON-NLS-2$
						+ ". holdCount:" + lock.getHoldCount() //$NON-NLS-1$
						+ ", queued:" + lock.getQueueLength()); //$NON-NLS-1$
			}
			return lock::unlock;
		}
		throw new ServiceException(
				"Failed to acquire lock within try period, Lock, id:" + System.identityHashCode(lock) + ", " //$NON-NLS-1$ //$NON-NLS-2$
						+ lock.toString(),
				ServiceException.DEADLOCK);
	}
```
What do you and the others think about using the `lock()` method proposed above. It would allow to use the 'locked-state' as resource of a try-block in a safe manner, at the three-locations where `use.getLock()` is now used:
```
try (Locked locked = use.lock()) {
 // perform the protected action
} catch (InterruptedException e) {
  Thread.currentThread().interrupt(); // To appease sonar
  // if necessary return a reasonable default
}
```
More or less `try (Locked locked = use.lock()) {` would simply replace `synchronized (use) {`.
This would make the `runWithLock()` methods obsolete since other exceptions can simply be propagated outside the try-block and would also simplify the lock's use in `ServiceRegistrationImpl.getService()`. Furthermore code duplications would be avoided (e.g. the 'DeadLock'-`ServiceException`).